### PR TITLE
1291 Date picker limits

### PR DESCRIPTION
**Pull request by IgorNadj**
### Changes:

- Add storybook stories for DateRangePicker
- Allow DateRangePicker to have custom limits set (not just 2015-today as is default, and this allows future limits too)
- Map overlay with date picker uses limits if specified
- Add tests for periodGranularities

---

### Screenshots:
https://github.com/beyondessential/tupaia-backlog/issues/1291#issuecomment-705304383

### Example configs:
E.g. limit 1 year ago, to yesterday:
```
# mapOverlay.presentationOptions
{
  "datePickerLimits": {
    "start": {
      "unit": "day",
      "offset": -365
    },
    "end": {
      "unit": "day",
      "offset": -1
    }
  },
  "scaleType": "neutral",
  "valueType": "oneDecimalPlace",
  "displayType": "spectrum",
  "measureLevel": "City",
  "defaultTimePeriod": {
    "unit": "day",
    "offset": -1
  },
  "periodGranularity": "one_day_at_a_time"
}
```
E.g. today to 16 days in the future:
```
# mapOverlay.presentationOptions
{
  "datePickerLimits": {
    "end": {
      "unit": "day",
      "offset": 16
    }
  },
  "scaleType": "neutral",
  "valueType": "oneDecimalPlace",
  "displayType": "spectrum",
  "measureLevel": "City",
  "periodGranularity": "one_day_at_a_time"
}
```

### CODE REVIEW

File: packages/web-frontend/src/utils/periodGranularities.js [hunk 1]
```diff
@@ -193,9 +203,16 @@ const getDefaultDatesForSingleDateGranularities = (periodGranularity, defaultTim
       singleDateConfig = defaultTimePeriod;
     }
 
+    const validDateOffsetUnit = GRANULARITIES_WITH_ONE_DATE_VALID_OFFSET_UNIT[periodGranularity];
+    if (singleDateConfig.unit !== validDateOffsetUnit) {
+      throw new Error(
```
Comment (IgorNadj) [inline packages/web-frontend/src/utils/periodGranularities.js:208]: I've checked the production dashboards and map overlays to see if any are configured with an offset unit that doesn't match the periodGranularity, and there are none. This will prevent devs from doing so in the future.

File: yarn.lock [hunk 2]
```diff
@@ -18,6 +18,13 @@
   optionalDependencies:
```
Comment (edmofro) [inline yarn.lock:18]: Seems like a lot of changes for one addition to package.json?

File: packages/web-frontend/src/components/DateRangePicker/DatePickerDialog.js [hunk 3]
```diff
@@ -135,7 +137,8 @@ export const DatePickerDialog = ({
     return setErrorMessage('');
   };
 
-  const minMomentDate = moment(MIN_DATE_PICKER_DATE);
+  const minMomentDate = min ? moment(min) : moment(MIN_DATE_PICKER_DATE);
```
Comment (edmofro) [inline packages/web-frontend/src/components/DateRangePicker/DatePickerDialog.js:140]:
```suggestion
  const minMomentDate = min ? moment(min) : moment(DEFAULT_MIN_DATE);
```
Worth a rename now it's just a default

File: packages/web-frontend/src/components/DateRangePicker/DateRangePicker.js [hunk 4]
```diff
@@ -116,25 +118,36 @@ export const DateRangePicker = ({
...
-  const defaultEndDate = isSingleDate ? defaultStartDate : maxMomentDate;
+  const minMomentDate = min ? moment(min) : moment(MIN_DATE_PICKER_DATE);
+  const maxMomentDate = max ? moment(max) : moment();
+
+  let defaultStartDate;
+  let defaultEndDate;
+  if (isSingleDate) {
+    defaultStartDate = moment();
+    if (defaultStartDate < minMomentDate) defaultStartDate = minMomentDate;
+    if (defaultStartDate > maxMomentDate) defaultStartDate = maxMomentDate;
```
Comment (edmofro) [inline packages/web-frontend/src/components/DateRangePicker/DateRangePicker.js:129]: Can you pull out a util called `constrainDate(defaultStartDate, minMomentDate, maxMomentDate)`? I reckon I'd find it more readable to have that as a named concept than as a list of conditions inline (especially as it's used in the next block too)

File: packages/web-frontend/src/containers/MeasureBar/Control.js [hunk 5]
```diff
@@ -110,6 +110,7 @@ export const Control = ({
   emptyMessage,
   selectedMeasure,
   defaultDates,
+  limits,
```
Comment (edmofro) [inline packages/web-frontend/src/containers/MeasureBar/Control.js:113]:
```suggestion
  datePickerLimits,
```
In the context of the measure bar control, I'd assume limits referred to constraints on the data itself

**Review by edmofro — COMMENTED** [review]
I'm finished for the day now so haven't done a full review, just had a couple of comments (hopefully someone else can review properly)

Nice work adding all those tests and dealing with the edge cases!!

File: yarn.lock (hunk 2 above)
Comment (IgorNadj) [reply yarn.lock:18]: Yeah it looks like a lot, seems the package I added has a lot of dev dependencies: https://github.com/Sambego/storybook-state/blob/master/package.json

I've had a look and it's the most popular for this purpose so I think we stay with it. Are we concerned at all with how big dev dependencies is getting?

File: packages/web-frontend/src/components/DateRangePicker/DatePickerDialog.js (hunk 3 above)
Comment (IgorNadj) [reply packages/web-frontend/src/components/DateRangePicker/DatePickerDialog.js]: Renamed

File: packages/web-frontend/src/components/DateRangePicker/DateRangePicker.js (hunk 4 above)
Comment (IgorNadj) [reply packages/web-frontend/src/components/DateRangePicker/DateRangePicker.js]: Extracted

File: packages/web-frontend/src/containers/MeasureBar/Control.js (hunk 5 above)
Comment (IgorNadj) [reply packages/web-frontend/src/containers/MeasureBar/Control.js]: Renamed

File: packages/web-frontend/src/tests/utils/periodGranularities.test.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+          },
+        });
+        expect(startDate.format()).toEqual('2019-02-04T00:00:00+11:00');
+        expect(endDate.format()).toEqual('2019-02-04T23:59:59+11:00');
+      });
+
+      it('throws if unit does not match period granularity', () => {
+        // The reason for this limitation is to prevent devs from confusing themselves.
+        // E.g. if periodGranularity is one_month_at_a_time, and offset is +5 days, it will not take
+        // effect, the start date will be rounded to the start of the month.
```
Comment (edmofro) [inline packages/web-frontend/src/tests/utils/periodGranularities.test.js:69]: Yeah great addition!

File: packages/web-frontend/src/tests/utils/periodGranularities.test.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+      // effect, the start date will be rounded to the start of the month.
+      const functionCall = () =>
+        getLimits('one_month_at_a_time', { start: { unit: 'day', offset: -3 } });
+      expect(functionCall).toThrow('limit unit must match periodGranularity');
+    });
+
+    it('calculates limits', () => {
+      const { startDate, endDate } = getLimits('day', { start: { unit: 'day', offset: -3 } });
+      expect(startDate.format()).toEqual(moment('2019-02-02T00:00:00').format()); // rounded
+      expect(endDate.format()).toEqual(moment('2019-02-05T23:59:59').format()); // rounded
```
Comment (edmofro) [inline packages/web-frontend/src/tests/utils/periodGranularities.test.js:166]: Can you do this, like above?
```suggestion
      expect(startDate.format()).toEqual('2019-02-02T00:00:00+11:00'); // rounded
      expect(endDate.format()).toEqual('2019-02-05T23:59:59+11:00'); // rounded
```

File: packages/web-frontend/src/tests/utils/periodGranularities.test.js [hunk 8]
```diff
@@ -0,0 +1,169 @@
+/**
+ * Tupaia Web
+ * Copyright (c) 2020 Beyond Essential Systems Pty Ltd.
+ * This source code is licensed under the AGPL-3.0 license
+ * found in the LICENSE file in the root directory of this source tree.
+ */
```
Comment (edmofro) [inline packages/web-frontend/src/tests/utils/periodGranularities.test.js:6]: Non-standard header

**Review by edmofro — APPROVED** [review]
Great tests!! Well needed in the period area, thanks for adding them

All approved, a couple of minor things added as comments earlier, but they're optional

File: packages/web-frontend/src/tests/utils/periodGranularities.test.js (hunk 7 above)
Comment (IgorNadj) [reply packages/web-frontend/src/tests/utils/periodGranularities.test.js]: Updated

File: packages/web-frontend/src/tests/utils/periodGranularities.test.js (hunk 8 above)
Comment (IgorNadj) [reply packages/web-frontend/src/tests/utils/periodGranularities.test.js]: Updated
